TESTS_COMMON/ExhaustibleBlockDevice.cpp

```diff
@@ -0,0 +1,203 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2017 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "ExhaustibleBlockDevice.h"
+
+
+static void program_data(uint8_t *dst, const uint8_t *src, bd_size_t size, bool prog_exhausted, bool erase_exhausted)
```

Comment:
If these were virtual, it would let an inheriting class override them. That way a user could just extend the ExhaustibleBlockDevice class to change how blocks die.

Actually I wonder if we could make HeapBlockDevice extendable this way, and just have ExhaustibleBlockDevice extend HeapBlockDevice with a `program_data` function that fails after n cycles...

Not sure it's relevant to this pr